trident/firefly_generator.py

```diff
@@ -0,0 +1,361 @@
+    #     extrema = 10 and 20, and dlambda (binsize) = 1, then n_lambda = 11.
+    #     Default: None
+
+    # :dlambda: float
+
+    #     size of the wavelength bins in angstroms or velocity bins in km/s.
+    #     Default: None
+
+    # :bin_space: 'wavelength' or 'velocity'
+
+    #     Sets the dimension in which spectra are created. If set to
+    #     wavelength, the resulting spectra are flux (or tau) vs. observed
+    #     wavelength. If set to velocity, the spectra are flux vs.
+    #     velocity offset from the rest wavelength of the absorption line.
+    #     Default: wavelength
+
+    # :lsf_kernel: string, optional
+
+    #     The filename for the LSF kernel. Files are found in
+    #     trident.__path__/data/lsf_kernels or current working directory.
+    #     Only necessary if you are applying an LSF to the spectrum in
+    #     postprocessing.
+    #     Default: None
+
+    # :line_database: string or :class:`~trident.LineDatabase`, optional
+
+    #     A text file listing the various lines to insert into the line database,
+    #     or a :class:`~trident.LineDatabase` object in memory. The line database
+    #     provides a list of all possible lines that could be added to the
+    #     spectrum. For a text file, it should have 4 tab-delimited columns of
+    #     name (e.g. MgII), wavelength in angstroms, gamma of transition, and
+    #     f-value of transition. See example datasets in
+    #     trident.path/data/line_lists for examples.
+    #     Default: lines.txt
+
+    # :ionization_table: hdf5 file, optional
+
+    #     An HDF5 file used for computing the ionization fraction of the gas
+    #     based on its density, temperature, metallicity, and redshift.  If
+    #     set to None, will use the ion table defined in your Trident config
+    #     file.
+    #     Default: None
+
+    # **Example**
+
+    # Create a one-zone ray, and generate a COS spectrum from that ray.
+
+    # >>> import trident
+    # >>> ray = trident.make_onezone_ray()
+    # >>> sg = trident.SpectrumGenerator('COS')
+    # >>> sg.make_spectrum(ray)
+    # >>> sg.plot_spectrum('spec_raw.png')
+
+    # Create a one-zone ray at redshift 0.5, and generate a spectrum with 1
+    # angstrom spectral bins from 2000-4000 angstroms, then post-process by
+    # adding a MW foreground a QSO background at z=0.5 and add a boxcar line
+    # spread function of 100 angstroms width.  Plot it and save the figure to
+    # 'spec_final.png'.
+
+    # >>> import trident
+    # >>> ray = trident.make_onezone_ray(redshift=0.5)
+    # >>> sg = trident.SpectrumGenerator(lambda_min=2000, lambda_max=4000,
+    # ... dlambda=1)
+    # >>> sg.make_spectrum(ray)
+    # >>> sg.add_qso_spectrum(emitting_redshift=.5)
+    # >>> sg.add_milky_way_foreground()
+    # >>> sg.apply_lsf(function='boxcar', width=100)
+    # >>> sg.plot_spectrum('spec_final.png')
+    """
+    def __init__(self, line_database='lines.txt',
```

Comment:
I'm guessing that this functionality will only work for particle-based datasets, right?  If so, perhaps we need a conditional somewhere at the beginning of this code that checks to see if this is particle-based or grid-based and raises a "Not Implemented error" for grid-based datasets?  Or something like that.